### PR TITLE
Add adapter modules

### LINKED ISSUE #325: Add support for per-adapter modules
This came up in the context of the aws adapter, where the current implementation did automatic aggregations of raw per-item information into aggregates (number of instances), demographics (breakdowns of instances by instance type), etc.

Although useful, it ends up hiding the raw per-item information and isn't in line with the data format returned by our other adapters.

One solution to this would be to define a module that created subgraphs that performed the aggregations (i.e. ec2_demographics() would be a | reduce count=count() by demographic). Juttle users who wanted these automatic rollups could just use the module. However, we don't have a great way to pass around modules in a way they can easily be used by all juttle users.

To address this, how about we allow adapters to publish modules along with their read and write procs? They could be namespaced as Adapter.<adapter_name>.*, and loaded into the juttle runtime when the adapter is loaded. This would allow us to provide useful aggregation methods, constants, etc along with each adapter that could make the raw data easier to use.

### PR DESCRIPTION
**Pull request by mstemm**
This adds support for per-adapter modules.

Should I bump the juttle api version for this? Following semver we'd want it to be 0.6.0, but maybe we could cheat and make it 0.5.3 instead.

After this merges I'll update the aws adapter and adapter template to provide examples.

This fixes #325.

@demmer @dmajda

### CODE REVIEW

File: lib/module-resolvers/file-resolver.js [hunk 1]
```diff
@@ -24,6 +27,22 @@ class FileResolver {
 
         this._additional_search_paths = options.search_paths || search_paths();
 
+        // Also add the module root for all adapters.
+        this._adapter_roots = [];
+
+        for(let adapter of adapters.list()) {
+            // Skip built-in adapters or adapters for which the
+            // install path is not absolute (i.e. not actually available)
+            if (adapter.module !== '(builtin)' &&
```
Comment (dmajda) [inline lib/module-resolvers/file-resolver.js]: Better check this with `adapter.builtin`.

File: lib/module-resolvers/file-resolver.js [hunk 2]
```diff
@@ -24,6 +27,22 @@ class FileResolver {
...
+                path.isAbsolute(adapter.path)) {
```
Comment (dmajda) [inline lib/module-resolvers/file-resolver.js]: I think it would be better to include a flag indicating if the adapter was actually loaded (e.g. `adapter.loaded`) and use it here. Checking the path is definitely not a clean check adapter’s state.

File: lib/module-resolvers/file-resolver.js [hunk 3]
```diff
@@ -24,6 +27,22 @@ class FileResolver {
...
+                let install_path = this._canonicalize(adapter.path);
+                this._adapter_roots.push({
+                    adapter: adapter.adapter,
+                    module_root: [path.dirname(install_path), ADAPTER_MODULE_SUBDIR].join(path.sep)
+                });
```
Comment (dmajda) [inline lib/module-resolvers/file-resolver.js]: Maybe add the path only when it exists and it is a directory? Otherwise we will unnecessarily check it on every module load.

File: lib/module-resolvers/file-resolver.js [hunk 4]
```diff
@@ -36,6 +55,23 @@ class FileResolver {
...
+            return {
+                source: fs.readFileSync(filename, 'utf8'),
+                name: (name ? name : filename)
+            };
+        }
+        catch (e) {
+            return false;
+        }
+    }
+
```
Comment (dmajda) [inline lib/module-resolvers/file-resolver.js]: The `_read_file` extraction would be better in a separate commit (it is orthogonal with the other changes).

File: lib/module-resolvers/file-resolver.js [hunk 5]
```diff
@@ -78,20 +114,30 @@ class FileResolver {
...
     _search_system_import(module_path, module_name, importer_path) {
         logger.debug(`searching system module_path=${module_path} module_name=${module_name} importer_path=${importer_path}`);
 
+        // First see if the module_path matches one of the adapter
+        // roots. If it does, prefix the adapter's module root, drop
+        // the adapter name from the module path, and try to read the
+        // file.
+        let parts = module_path.split(path.sep);
+        let first_subdir = parts.shift();
+        for(let root of this._adapter_roots) {
```
Comment (dmajda) [inline lib/module-resolvers/file-resolver.js]: Use `_.findWhere` instead of a `for` loop?

**Comment by dmajda** [comment]
Beside the minor comments inline, I have two bigger notes:

  1. I think we should prefix adapter module paths with a common prefix, e.g. `adapters`. So instead of:

    ```juttle
    import "aws/aggregations" as Adapter_aws;
    ```

   one would do:

    ```juttle
    import "adapters/aws/aggregations" as Adapter_aws;
    ```

    My main concern is that I can imagine stdlib growing subdirectories at some point and it wouldn’t be good to thare that namespace with adapters (which would then be able to shadow stdlib stuff or vice versa).

  2. I don’t like the `module_root` directory name, it wouldn’t say anything to me if I saw adapter code for the first time. I’d prefer `juttle`, `juttle-modules`, or `modules` (in that order); there may be some other good names too.

@demmer I’d like to have your opinion on both points.

**Comment by dmajda** [comment]
> Should I bump the juttle api version for this? Following semver we'd want it to be 0.6.0, but maybe we could cheat and make it 0.5.3 instead.

I think so (we are providing a new capability), I’d go for 0.6.0 (I don’t see a reason for “cheating”).

**Comment by mstemm** [comment]
I like the top-level adapters directory, I'll do that.

@demmer didn't like module_root either but didn't have any good alternatives. I like most of yours better. He also wanted the name juttle in the name somewhere so I think I'll use juttle-modules.

**Comment by mstemm** [comment]
My understanding of semver for versions < 1.0 is that minor number bumps are not considered compatible with lower minor numbers. So if we add this (really backwards-compatible) feature and increment the minor version we'd have to bump the adapter api version of all adapters.

**Comment by demmer** [comment]
>> I don’t like the module_root directory name, it wouldn’t say anything to me if I saw adapter code for the first time. I’d prefer juttle, juttle-modules, or modules (in that order); there may be some other good names too.

> @demmer didn't like module_root either but didn't have any good alternatives. I like most of yours better. He also wanted the name juttle in the name somewhere so I think I'll use juttle-modules.

I can live with `juttle-modules` but I'd slightly prefer the simpler
`juttle` or `juttle-lib`.


On Fri, Mar 18, 2016 at 9:32 AM, Mark Stemm <notifications@github.com> wrote:
> I like the top-level adapters directory, I'll do that.
>
> @demmer didn't like module_root either but didn't have any good
> alternatives. I like most of yours better. He also wanted the name juttle in
> the name somewhere so I think I'll use juttle-modules.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub



-- 
Michael Demmer
Jut - CTO / VP Engineering
demmer@jut.io

**Comment by mstemm** [comment]
I don't like plain 'juttle' but it's at the top of both of your lists and I want this done, so I'll switch to juttle.

**Comment by dmajda** [comment]
> My understanding of semver for versions < 1.0 is that minor number bumps are not considered compatible with lower minor numbers. So if we add this (really backwards-compatible) feature and increment the minor version we'd have to bump the adapter api version of all adapters.

One thing is what the semver spec itself says, which is basically [“all bets are off in < 1.0.0”](http://semver.org/#spec-item-4). From that standpoint, we shouldn’t use version ranges in adapters’ `package.json` at all.

Another thing is what the `^` operator does in Node.js (or, more precisely, in `semver.satisifes`). This is basically what you say and yes, we’d need to bump if we’d go to 0.6.0 (or 0.7.0 to correspond to the next release).

Reading the [`^` docs](https://github.com/npm/node-semver#caret-ranges-123-025-004), it looks like the idea is basically that the version semantics is shifted by one step in the < 1.0.0. This would correspond to your 0.5.3 suggestion.

Considering all this, I think you are right and we should just go with 0.5.3 to avoid the hassle of updating all adapters.

File: lib/module-resolvers/file-resolver.js [hunk 6]
```diff
@@ -36,6 +36,23 @@ class FileResolver {
         }
     }
 
+    _read_file(filename, name) {
+
```
Comment (dmajda) [inline lib/module-resolvers/file-resolver.js]: Nit: Omit the empty line?

File: lib/module-resolvers/file-resolver.js [hunk 7]
```diff
@@ -24,6 +28,29 @@ class FileResolver {
 
         this._additional_search_paths = options.search_paths || search_paths();
 
+        // Also add the module root for all adapters.
+        this._adapter_roots = [];
+
+        for(let adapter of adapters.list()) {
```
Comment (dmajda) [inline lib/module-resolvers/file-resolver.js]: Nit: Space after `for`.

File: lib/module-resolvers/file-resolver.js [hunk 8]
```diff
@@ -24,6 +28,29 @@ class FileResolver {
...
+            // Only consider non built-in adapters that were loaded successfully.
+            if (! adapter.builtin &&
+                adapter.loaded) {
```
Comment (dmajda) [inline lib/module-resolvers/file-resolver.js]: Nit: I think the condition can just be on one line.

File: lib/module-resolvers/file-resolver.js [hunk 9]
```diff
@@ -99,9 +126,40 @@ class FileResolver {
         return this._read_file(filename);
     }
 
+    _search_adapter_import(module_path, module_name, importer_path) {
+
+        // Remove the top 'adapters' directory
+        let parts = module_path.split(path.sep).slice(1);
```
Comment (dmajda) [inline lib/module-resolvers/file-resolver.js]: Hmm, this makes Juttles incompatible between Unix and Windows because one would have to use `\` as a separator on Windows. I guess we just need to hard-code `/` here.

File: lib/module-resolvers/file-resolver.js [hunk 10]
```diff
@@ -99,9 +126,40 @@ class FileResolver {
...
+
+        return false;
+    }
+
     _search_system_import(module_path, module_name, importer_path) {
         logger.debug(`searching system module_path=${module_path} module_name=${module_name} importer_path=${importer_path}`);
 
+        // If the module_path starts with 'adapters' + pathsep, assume
+        // it's an adapter module load.
+        if (module_path.startsWith('adapters' + path.sep)) {
```
Comment (dmajda) [inline lib/module-resolvers/file-resolver.js]: Same here (hard-code `/`).

**Comment by dmajda** [comment]
+1 with the comments addressed (and sorry I didn’t catch some of these during the first round)
